src/hooks/useSwapCallback.ts

```diff
@@ -337,51 +377,81 @@ export function useSwapCallback(
             : {}),
           ...(value && !isZero(value) ? { value } : {}),
         }
-
-        if (frontrunningProtection && isMetaMask) {
-          const serialized = serializeTransaction(transaction)
+        if (frontrunningProtection && isMetaMask && chainId === 1) {
+          let web3Provider: Web3Provider | undefined
+          // ethers will change eth_sign to personal_sign if it detects metamask
+          if (library instanceof Web3Provider) {
+            web3Provider = library as Web3Provider
+            isMetaMask = web3Provider.provider.isMetaMask
+            web3Provider.provider.isMetaMask = false
+          }
+          const nonce = await library.getTransactionCount(recipient)
+          const populatedTx: PopulatedTransaction = {
+            to: transaction.to,
+            nonce,
+            chainId,
+            type: 2,
+            data: transaction.data,
+            gasLimit: transaction.gasLimit,
+            maxFeePerGas: feesPerGas.maxFeePerGas,
+            maxPriorityFeePerGas: feesPerGas.maxPriorityFeePerGas,
+            ...(value && !isZero(value) ? { value: BigNumber.from(value) } : {}),
+          }
+          delete populatedTx.from
+          const serialized = serializeTransaction(populatedTx)
           const hash = keccak256(serialized)
-
           return library
             .jsonRpcFetchFunc('eth_sign', [account, hash])
-            .then((signature: SignatureLike) => serializeTransaction(transaction, signature))
+            .then((signature: SignatureLike) => serializeTransaction(populatedTx, signature))
             .catch((error) => {
               // otherwise, the error was unexpected and we need to convey that
               console.error(`Failed to sign transaction`, error, address, calldata, value)
-
+              if (web3Provider) {
+                web3Provider.provider.isMetaMask = true
+              }
               throw new Error(`Swap failed: ${swapErrorToUserReadableMessage(error)}`)
             })
             .then((rawSignedTransaction) => {
-              return rawSignedTransaction
+              if (web3Provider) {
+                web3Provider.provider.isMetaMask = true
+              }
+
+              /**
+               * Hack for detecting hardware wallets connected through metamask
+               * Hardware wallets cannot sign with eth_sign. A direct hardware wallet connection should be used instead
+               */
+              const parsed = parseTransaction(rawSignedTransaction)
+              if (parsed.from !== account) {
```

Comment:
@Dire-0x 
Two things I wanted to mention
1. The 'from' address could differ on non supported wallets too - not just hardware, so maybe a copy that covers both scenarios?
2. A secondary check could also be done to check the nonce number against the `getTransactionCount` function to further identify other incompatible wallets.